### ISSUE TITLE
Pywr is incompatible with cython 3.0.0

### ISSUE THREAD

**Issue by knoxsp** [pywr/pywr#1102]
Trying to install Pywr using cython 3.0.0 (the latest as of this date) results in the error:
```
[ 1/10] Cythonizing pywr/solvers/cython_glpk.pyx

Error compiling Cython file:
------------------------------------------------------------
...
    def __init__(self, use_presolve=False, time_limit=None, iteration_limit=None, message_level="error",
                 set_fixed_flows_once=False, set_fixed_costs_once=False, set_fixed_factors_once=False, use_unsafe_api=False):
        self.use_presolve = use_presolve
        self.use_unsafe_api = use_unsafe_api
        self.set_solver_options(time_limit, iteration_limit, message_level)
        glp_term_hook(term_hook, NULL)
                      ^
------------------------------------------------------------

pywr/solvers/cython_glpk.pyx:158:22: Cannot assign type 'int (void *, const char *) except? -1' to 'int (*)(void *, const char *) noexcept'

Error compiling Cython file:
------------------------------------------------------------
...
        self.has_presolved = False
        self.set_fixed_flows_once = set_fixed_flows_once
        self.set_fixed_costs_once = set_fixed_costs_once
        self.set_fixed_factors_once = set_fixed_factors_once
        # Register the error hook
        glp_error_hook(error_hook, &error_ctx)
                       ^
------------------------------------------------------------
```

**Comment by knoxsp** [comment]
Reverting cython to 0.29.36 resolves this issue

**Comment by jetuk** [comment]
We pinned the version to < 3 in `pyproject.toml` in the last release. 

The issue is to do with handling exceptions in the Cython. I've not got around to digging in to it. There's a migration guide here: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html

**Comment by knoxsp** [comment]
Thanks for the response. I mainly wanted this issue open in case others encouter the same problem and searched for a solution

**Comment by jetuk** [comment]
Fixed in #1104
